src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -1614,11 +1621,8 @@ size_t ObjectSynchronizer::deflate_idle_monitors() {
   size_t deflated_count = deflate_monitor_list(current, ls, &timer);
   size_t unlinked_count = 0;
   size_t deleted_count = 0;
-  if (deflated_count > 0 || is_final_audit()) {
-    // There are ObjectMonitors that have been deflated or this is the
-    // final audit and all the remaining ObjectMonitors have been
-    // deflated, BUT the MonitorDeflationThread blocked for the final
-    // safepoint during unlinking.
+  if (deflated_count > 0) {
```

Comment:
There's a conditional at line 1631 is_monitor_deflation_thread, and an else that deflating idle monitors is at a safepoint.  Should there not be this conditional, ie am I reading the description right?

Reply:
You are right. The plan in my head has been to remove the safepointed monitor deflation first and then take a pass over the code and remove the various blocks that we don't execute anymore.

Reply:
@coleenp I created a PR to clean this up:
https://github.com/openjdk/jdk/pull/16706